include/sys/zio_crypt.h

```diff
@@ -0,0 +1,157 @@
+/*
+ * CDDL HEADER START
+ *
+ * This file and its contents are supplied under the terms of the
+ * Common Development and Distribution License ("CDDL"), version 1.0.
+ * You may only use this file in accordance with the terms of version
+ * 1.0 of the CDDL.
+ *
+ * A full copy of the text of the CDDL should have accompanied this
+ * source.  A copy of the CDDL is also available via the Internet at
+ * http://www.illumos.org/license/CDDL.
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2016, Datto, Inc. All rights reserved.
+ */
+
+#ifndef	_SYS_ZIO_CRYPT_H
+#define	_SYS_ZIO_CRYPT_H
+
+#include <sys/dmu.h>
+#include <sys/refcount.h>
+#include <sys/crypto/api.h>
+#include <sys/nvpair.h>
+#include <sys/avl.h>
+#include <sys/zio.h>
+
+/* forward declarations */
+struct zbookmark_phys;
+
+#define	WRAPPING_KEY_LEN 32
+#define	WRAPPING_IV_LEN DATA_IV_LEN
+#define	WRAPPING_MAC_LEN 16
+
+#define	SHA1_DIGEST_LEN 20
+#define	SHA_256_DIGEST_LEN 32
+#define	HMAC_SHA256_KEYLEN 32
+
+#define	L2ARC_DEFAULT_CRYPT ZIO_CRYPT_AES_256_CCM
+
+#define	ZIO_NO_ENCRYPTION_NEEDED -1
+
+/*
+ * After encrypting many blocks with the same salt we may start to run
+ * up against the theoretical limits of how much data can securely be
+ * encrypted a single key using the supported encryption modes. To
+ * counteract this we generate a new salt after writing
+ * ZIO_CRYPT_MAX_SALT_USAGE blocks of data, tracked by zk_salt_count.
+ * The current value was chosen because it is approximately the number
+ * of blocks that would have to be written in order to acheive a
+ * 1 / 1 trillion chance of having an IV collision. Developers looking to
+ * change this number should make sure they take into account the
+ * birthday problem in regards to IV generation and the limits of what the
+ * underlying mode can actually handle.
+ */
+#define	ZIO_CRYPT_MAX_SALT_USAGE 400000000
+
+/* utility macros */
+#define	BITS_TO_BYTES(x) (((x) + 7) >> 3)
+#define	BYTES_TO_BITS(x) (x << 3)
+
+typedef enum zio_crypt_type {
+	ZC_TYPE_NONE = 0,
+	ZC_TYPE_CCM,
+	ZC_TYPE_GCM
+} zio_crypt_type_t;
+
+/* table of supported crypto algorithms, modes and keylengths. */
+typedef struct zio_crypt_info {
+	/* mechanism name, needed by ICP */
+	crypto_mech_name_t ci_mechname;
+
+	/* cipher mode type (GCM, CCM) */
+	zio_crypt_type_t ci_crypt_type;
+
+	/* length of the encryption key */
+	size_t ci_keylen;
+
+	/* human-readable name of the encryption alforithm */
+	char *ci_name;
+} zio_crypt_info_t;
+
+extern zio_crypt_info_t zio_crypt_table[ZIO_CRYPT_FUNCTIONS];
+
+/* ZAP entry keys for DSL Encryption Keys stored on disk */
+#define	DSL_CRYPTO_KEY_CRYPT "DSL_CRYPTO_CRYPT"
+#define	DSL_CRYPTO_KEY_IV "DSL_CRYPTO_IV"
+#define	DSL_CRYPTO_KEY_MAC "DSL_CRYPTO_MAC"
+#define	DSL_CRYPTO_KEY_MASTER_BUF "DSL_CRYPTO_MASTER"
+#define	DSL_CRYPTO_KEY_HMAC_KEY_BUF "DSL_CRYPTO_HMAC_KEY"
```

Comment:
Any reason for the macro names and values to not be consistent, at least `DSL_CRYPTO_KEY_XXX -> "DSL_CRYPTO_XXX"`  (this applies to the last 2 which differ in the `_BUF` suffix)